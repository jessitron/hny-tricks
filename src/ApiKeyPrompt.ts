import { Html, html } from "./htm-but-right";
import { trace } from "@opentelemetry/api";
import {
  Region,
  KeyType,
  KeyInfo,
  EnvironmentType,
  HoneycombAuthResponse,
  HoneycombApiEndpointByRegion,
  describeAuthorization,
  HnyTricksAuthorization,
} from "./common";
import { fetchFromHoneycombApi, isFetchError } from "./HoneycombApi";
import {
  currentTraceId,
  inSpan,
  inSpanAsync,
  recordError,
  report,
} from "./tracing-util";

export const ELEMENT_CONTAINING_ALL_SECTIONS = "#stuff";
export const ENDPOINT_TO_POPULATE_ALL_SECTIONS = "/team";

export const startingApiKeyPrompt = html`<${ApiKeyPrompt}
  destinationElement="${ELEMENT_CONTAINING_ALL_SECTIONS}"
  endpointToPopulateItWith="${ENDPOINT_TO_POPULATE_ALL_SECTIONS}"
  endpointForApiKeyValidation="/validate"
/>`;

export function ApiKeyPrompt(params: {
  destinationElement: string;
  endpointToPopulateItWith: string;
  endpointForApiKeyValidation: string;
}): Html {
  return html`<section id="apikey-section" class="apiKey">
    <form hx-post="${params.endpointToPopulateItWith}" hx-target="${params.destinationElement}" hx-swap="innerHTML" id="apikey-form" hx-indicator="#big-think">
      <div>
        <label for="apikey">Honeycomb API Key <a href="https://docs.honeycomb.io/get-started/configure/environments/manage-api-keys/" target="_blank">
         ‚ùî
          </a> <span id="reveal-password" tron-reveal="#apikey">üëÅ</span></label>
        <input id="apikey" type="password" 
          name="apikey"
          hx-trigger="input changed throttle:200ms"
          hx-post="${params.endpointForApiKeyValidation}"
          hx-target="#apikey-opinion"
          hx-include="#apikey" 
          hx-swap="innerHTML"></input>
        <button>Give Me Options</button>
      </div>
      <div id="apikey-opinion">
      </div>
    </form>
    <script src="/jess.js"></script>
 </section>`;
}

export function commentOnApiKey(apiKey: string): Html {
  if (!apiKey) {
    return "";
  }
  const keyInfo = interpretApiKey(apiKey);
  const remark = remarkOnKeyInfo(keyInfo);

  return html`<span
    class="${remark.className}"
    data-traceid="${currentTraceId()}"
    >${remark.description}</span
  >`;
}

function remarkOnKeyInfo(keyInfo): {
  className: "happy" | "unhappy";
  description: string;
} {
  if (keyInfo.type === "ingest") {
    if (keyInfo.region !== "unknown") {
      return {
        className: "happy",
        description: `This looks like a ${keyInfo.region} ingest key`,
      };
    } else {
      return {
        className: "unhappy",
        description:
          "This looks like an ingest key, but I can't tell which region",
      };
    }
  } else if (
    keyInfo.type === "configuration" &&
    keyInfo.environmentType === "classic"
  ) {
    return {
      className: "happy",
      description: "That looks like a Honeycomb Classic configuration key.",
    };
  } else if (keyInfo.type === "configuration") {
    return {
      className: "happy",
      description: "That looks like a Honeycomb configuration key. Great.",
    };
  }
  return {
    className: "unhappy",
    description: "That doesn't look like an API key",
  };
}

export function regionByLetter(letter: string): Region {
  switch (letter) {
    case "a":
      return "US";
    case "b":
      return "EU";
    case "c":
      return "dogfood US";
    case "d":
      return "dogfood EU";
    case "e":
      return "kibble US";
    case "f":
      return "kibble EU";
  }
}

export function interpretApiKey(apiKey: string): KeyInfo {
  let keyType: KeyType = "none";
  let region: Region = "unknown";
  let environmentType: EnvironmentType = "none";
  if (apiKey.length === 64 && apiKey.match(/^hc[abcdef]i[kc]_[a-z0-9]{58}$/)) {
    keyType = "ingest";
    region = regionByLetter(apiKey[2]);
    switch (apiKey[4]) {
      case "c":
        environmentType = "classic";
        break;
      case "k":
        environmentType = "e&s";
        break;
    }
  } else if (apiKey.match(/^[a-zA-Z0-9]{22}$/)) {
    keyType = "configuration";
    environmentType = "e&s";
    region = "unknowable";
  } else if (apiKey.match(/^[a-f0-9]{32}$/)) {
    keyType = "configuration";
    environmentType = "classic";
    region = "unknowable";
  } else if (apiKey.match(/^hc[abcdef]mk_[a-z0-9]{26}$/)) {
    keyType = "management key ID";
    environmentType = "none";
    region = regionByLetter(apiKey[2]);
  } else if (apiKey.match(/^[a-z0-9]{32}$/)) {
    keyType = "management key secret";
    environmentType = "none";
    region = "unknowable";
  }
  trace.getActiveSpan()?.setAttributes({
    "honeycomb.key.type": keyType,
    "honeycomb.key.region": region,
    "honeycomb.key.environmentType": environmentType,
  });
  return { type: keyType, environmentType, region };
}

export type AuthError = {
  authError: true;
  html: Html;
};
function authError(htmlInput: Html): AuthError {
  const wrappedHtml = html`<div data-traceid="${currentTraceId()}">
    ${htmlInput}
  </div>`;

  return { authError: true, html: wrappedHtml };
}
export function isAuthError(
  response: HnyTricksAuthorization | AuthError
): response is AuthError {
  return (response as AuthError).authError;
}

export async function authorize(
  apiKey: string
): Promise<HnyTricksAuthorization | AuthError> {
  if (!apiKey) {
    recordError(new Error("no API key provided"));
    return authError(
      html`<div>
        <span class="unhappy">
          Trying to authorize, but API key is blank. This part doesn't work yet
        </span>
      </div>`
    );
  }

  const keyInfo = interpretApiKey(apiKey);
  if (keyInfo.region === "unknown") {
    return authError(html`<div>
      <span class="unhappy">This doesn't look like an API key</span>
    </div>`);
  }
  if (keyInfo.region === "unknowable") {
    const workingRegion = await tryAllRegions(apiKey);
    if (workingRegion === "unknown") {
      return authError(html`<div>
        <span class="unhappy">
          That API key did not work in any Honeycomb region. üò≠
        </span>
      </div>`);
    }
    keyInfo.region = workingRegion; // now we know.
  }
  const response = await fetchFromHoneycombApi<HoneycombAuthResponse>(
    { apiKey, keyInfo },
    "auth"
  );
  trace
    .getActiveSpan()
    ?.setAttributes({ "honeycomb.auth.response": JSON.stringify(response) });
  if (isFetchError(response)) {
    return authError(html`<div>
      <span class="unhappy">Auth check failed: ${response.message}</span>
    </div>`);
  }
  return describeAuthorization(apiKey, keyInfo, response);
}

async function tryAllRegions(apiKey: string): Promise<Region> {
  return inSpanAsync("try the API key in all regions", async (span) => {
    const regions = Object.keys(HoneycombApiEndpointByRegion) as Region[]; // is there a cleverer way to do that?
    const regionIdentified = (await Promise.any(
      regions.map((region) =>
        inSpanAsync("try region " + region, async () => {
          const response = await fetchFromHoneycombApi<HoneycombAuthResponse>(
            { apiKey, keyInfo: { region } },
            "auth"
          );
          if (isFetchError(response)) {
            throw new Error("Well, it didn't work in " + region);
          } else {
            return region;
          }
        })
      )
    ).catch((allRejected) => {
      return "unknown";
    })) as Region;

    report({
      "honeycomb.region": regionIdentified,
      "app.regionsTried": regions.length,
    });
    return regionIdentified;
  });
}
