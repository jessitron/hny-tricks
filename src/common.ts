export type HoneycombAuthResponse = {
  id: string;
  type: "configuration" | "ingest";
  api_key_access: {
    events: boolean;
    markers: boolean;
    triggers: boolean;
    boards: boolean;
    queries: boolean;
    columns: boolean;
    createDatasets: boolean;
    slos: boolean;
    recipients: boolean;
    privateBoards: boolean;
  };
  environment: {
    name: string;
    slug: string;
  };
  team: {
    name: string;
    slug: string;
  };
};

export type KeyType = "none" | "ingest" | "configuration";
export type Region =
  | "US"
  | "EU"
  | "dogfood EU"
  | "dogfood US"
  | "unknown"
  | "unknowable"; // configuration keys don't include region info in the key
export type EnvironmentType = "classic" | "e&s" | "none";
export type KeyInfo = {
  type: KeyType;
  environmentType: EnvironmentType;
  region: Region;
};

// @ts-ignore I don't want this to be complete, TypeScript. I do want all the keys to be regions.
export const HoneycombEndpointByRegion: Record<Region, string> = {
  EU: "https://api.eu1.honeycomb.io/1/",
  US: "https://api.honeycomb.io/1/",
  "dogfood EU": "https://api.dogfood.eu1.honeycomb.io/1/",
  "dogfood US": "https://api-dogfood.honeycomb.io/1/",
};

// @ts-ignore I don't want this to be complete, TypeScript. I do want all the keys to be regions.
export const HoneycombUIEndpointByRegion: Record<Region, string> = {
  EU: "https://ui.eu1.honeycomb.io/",
  US: "https://ui.honeycomb.io/",
  "dogfood EU": "https://ui.dogfood.eu1.honeycomb.io/",
  "dogfood US": "https://ui-dogfood.honeycomb.io/",
};
