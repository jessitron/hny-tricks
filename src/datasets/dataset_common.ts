
export type DatasetSlug = string;

export type HnyTricksDataset = {
  name: string;
  slug: DatasetSlug;
  created: Date;
  last_written: Date;
};


export type Html = string;
export interface Column {
  header(): Html;
  row(d: HnyTricksDataset, i: number): Html;
  footer(): Html;
}
