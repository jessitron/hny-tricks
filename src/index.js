"use strict";
const greet = "Hello wrold";
console.log(greet);
