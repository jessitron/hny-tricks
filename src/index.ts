import "./tracing";
import express, { Request, Response } from "express";
import { html } from "./htm-but-right";
import { trace } from "@opentelemetry/api";
import { Team as HoneycombApiKey } from "./HoneycombApiKey";
import bodyParser from "body-parser";

const app = express();
const port = process.env.PORT || 3000;

// serve files from the public directory
app.use(express.static("public"));
app.use(bodyParser.urlencoded({ extended: true }));

app.listen(port, () => {
  console.log(`http://localhost:${port}`);
});

app.get("/", (req, res) => {
  console.log("here we are at /");
  // well, I found the trick to get it to set the script body in the script :-/
  const element = html`<html>
    <head>
      <script src="https://unpkg.com/@jessitronica/hny-otel-web@0.2.0/dist/hny.js"></script>
      <script
        dangerouslySetInnerHTML=${{ __html: javascriptToStartTracing }}
      ></script>
      <script src="/htmx.js"></script>
      <title>Hny Tricks</title>
      <link rel="stylesheet" href="styles.css" />
    </head>
    <body>
      <h1>Jessitron's Honeycomb Tricks</h1>
      <${HoneycombApiKey} />
      <div id="stuff"></div>
    </body>
  </html>`;
  const fullResponse = `<!DOCTYPE html>${element}`;
  trace.getActiveSpan()?.setAttribute("response.body", fullResponse);
  res.send(fullResponse);
});

app.post("/team", (req: Request, res: Response) => {
  console.log("here we are at /team");
  res.send("ok");
});

app.post("/validate", (req: Request, res: Response) => {
  console.log("here we are at /validate oy");
  trace.getActiveSpan()?.setAttributes({
    dammit: "work",
    "noreally.request.body": "<" + JSON.stringify(req.body) + ">",
    "noreally.request.query": "<" + JSON.stringify(req.query) + ">",
    "noreally.request.params": "<" + JSON.stringify(req.params) + ">",
  });
  res.send("ðŸ˜µ");
});

const javascriptToStartTracing = `
  console.log("I am the script yo");
   Hny({  apiKey: "hcaik_01j229z3ezkranjdcgj7e5dvx4vz4prbbkkbgx4rnv4strm0kb485sae9f",
          serviceName: "hny-tricks-web",
          debug: true,
        });
`;
