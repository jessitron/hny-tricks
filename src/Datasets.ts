import { HnyTricksAuthorization } from "./common";
import { html } from "./htm-but-right";

export function describeDatasets(auth: HnyTricksAuthorization): string {
  return html`<section>Datasets Go Here</section>`;
}
