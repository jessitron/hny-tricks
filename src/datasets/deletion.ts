import { trace } from "@opentelemetry/api";
import { HnyTricksAuthorization } from "../common";
import { fetchFromHoneycombApi, isFetchError } from "../HoneycombApi";
import { Html, html } from "../htm-but-right";
import { inSpanAsync } from "../tracing-util";
import { Column, DatasetSlug, HnyTricksDataset } from "./dataset_common";
import { StatusUpdate } from "../status";

const ASSUMED_RETENTION_TIME = 60; // days

export class DaysSinceLastWritten implements Column {
  private minDaysSinceLastWrite: number;
  private now = new Date();
  private daysSince(date: Date) {
    return Math.floor(
      (this.now.getTime() - date.getTime()) / (1000 * 60 * 60 * 24)
    );
  }
  constructor(allLastWrittenDates: Date[]) {
    this.minDaysSinceLastWrite = Math.min(
      ...allLastWrittenDates.map((a) => this.daysSince(a))
    );
  }

  header(): Html {
    return html`<th scope="col">Days Since Last Data</th>`;
  }
  row(d: HnyTricksDataset): Html {
    return html`<td>${"" + this.daysSince(d.last_written)}</td>`;
  }
  footer(): Html {
    return html`<td>${"" + this.minDaysSinceLastWrite}</td>`;
  }
}

export class DeleteMe implements Column {
  private now = new Date();
  private daysSince(date: Date) {
    return Math.floor(
      (this.now.getTime() - date.getTime()) / (1000 * 60 * 60 * 24)
    );
  }
  constructor() {}

  header(): Html {
    return html`<th scope="col">Delete?</th>`;
  }
  row(d: HnyTricksDataset, i: number): Html {
    const slug = html`<input
      name="dataset-slug-${i}"
      value="${d.slug}"
      type="hidden"
    />`;
    const checkbox =
      this.daysSince(d.last_written) > ASSUMED_RETENTION_TIME
        ? html`<input
            class="delete-dataset-checkbox"
            type="checkbox"
            name="delete-dataset-${i}"
            checked
          />`
        : html`<input
            class="delete-dataset-checkbox"
            type="checkbox"
            title="I only delete empty datasets"
            disabled
          />`; // don't delete datasets with data in them
    return html`<td>${slug}${checkbox}</td>`;
  }
  footer(): Html {
    return html`<td>
      <button
        hx-post="/datasets/delete"
        hx-target="#dataset-section"
        hx-include="#auth_data"
      >
        Delete Old Datasets
      </button>
    </td>`;
  }
}

/**
 * for indexes i...
 * {
 * delete-dataset-${i} : "on" // or else it's absent
 * dataset-slug-${i}: string
 * }
 */
export type DeleteDatasetInputs = Record<string, string>;
export async function deleteDatasets(
  auth: HnyTricksAuthorization,
  inputs: DeleteDatasetInputs
): Promise<StatusUpdate> {
  const span = trace.getActiveSpan();
  span?.setAttributes({ "app.datasets.delete.input": JSON.stringify(inputs) });
  const datasetSlugs = datasetSlugsToDelete(inputs);

  const results = await Promise.all(
    datasetSlugs.map((slug) =>
      inSpanAsync("delete dataset " + slug, () =>
        enableDatasetDeletion(auth, slug).then((status) =>
          deleteDataset(auth, status)
        )
      )
    )
  );

  const success = results.every((r) => r.deleted);

  const status =
    results.length === 0
      ? html`Zero datasets deleted`
      : results.map((r) => {
          if (r.deleted === true) {
            return html`<p class="success-result">${r.slug} deleted ðŸ™‚</p>`;
          }
          return html`<p class="failure-result">
            ${r.slug} not deleted: ${r.error} ðŸ˜­
          </p>`;
        });

  return { success, html: status };
}

async function enableDatasetDeletion(
  auth: HnyTricksAuthorization,
  slug: DatasetSlug
): Promise<DatasetDeletionStatus> {
  // the [docs](https://api-docs.honeycomb.io/api/datasets/updatedataset) for this say that I should
  // send all fields or else they'll be wiped out (like the description) but wth we're deleting it anyway
  const newDatasetSettings = {
    settings: {
      delete_protected: false,
    },
  };
  const result = await fetchFromHoneycombApi(
    {
      apiKey: auth.apiKey,
      method: "PUT",
      keyInfo: auth.keyInfo,
      body: newDatasetSettings,
      whatToExpectBack: "nothing",
    },
    "datasets/" + slug
  );
  if (isFetchError(result)) {
    return {
      slug,
      deletionEnabled: false,
      deleted: false,
      error: result.message,
    };
  }
  return { slug, deletionEnabled: true, deleted: false };
}

type DatasetDeletionStatus = {
  slug: DatasetSlug;
  deleted: boolean;
  deletionEnabled: boolean;
  error?: string;
};

async function deleteDataset(
  auth: HnyTricksAuthorization,
  status: DatasetDeletionStatus
): Promise<DatasetDeletionStatus> {
  const { slug } = status;
  if (!status.deletionEnabled) {
    return status;
  }

  const result = await fetchFromHoneycombApi(
    {
      apiKey: auth.apiKey,
      method: "DELETE",
      keyInfo: auth.keyInfo,
      whatToExpectBack: "nothing",
    },
    "datasets/" + slug
  );
  if (isFetchError(result)) {
    return { ...status, deleted: false, error: result.message };
  }
  return { ...status, deleted: true };
}

// exported for testing
export function datasetSlugsToDelete(
  inputs: DeleteDatasetInputs
): DatasetSlug[] {
  const datasetSlugs = Object.keys(inputs)
    .filter((k) => k.startsWith("delete-dataset-"))
    .map((k) => k.split("-"))
    .map((parts) => parts[2]) // get just the number
    .map((i) => inputs[`dataset-slug-${i}`]); // get the slug for that number
  return datasetSlugs;
}
