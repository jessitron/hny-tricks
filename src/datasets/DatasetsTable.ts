import { HnyTricksAuthorization, constructEnvironmentLink } from "../common";
import { html } from "../htm-but-right";
import { Column, HnyTricksDataset, Html } from "./dataset_common";
import { DaysSinceLastWritten, DeleteMe } from "./deletion";
import { DerivedColumnForDatasetName } from "./derivedColumns";

export function DatasetsTable(params: {
  datasets: HnyTricksDataset[];
  auth: HnyTricksAuthorization;
}) {
  const { datasets, auth } = params;

  const environmentUrl = constructEnvironmentLink(auth);
  const columns = [
    new DatasetName(datasets.length, auth.environment.name),
    new LinkToSettings(environmentUrl),
    new LinkToQuery(environmentUrl),
    new DaysSinceLastWritten(datasets.map((d) => d.last_written)),
    new DeleteMe(),
    new DerivedColumnForDatasetName(),
  ];
  return html`<form id="dataset-table-form">
    <table class="dataset-table">
      <thead>
        <tr>
          ${columns.map((c) => c.header())}
        </tr>
      </thead>
      ${datasets.map(
        (d, i) =>
          html`<tr>
            ${columns.map((c) => c.row(d, i))}
          </tr>`
      )}
      <tfoot>
        <tr>
          ${columns.map((c) => c.footer())}
        </tr>
      </tfoot>
    </table>
  </form>`;
}

class DatasetName implements Column {
  constructor(
    private countOfDatasets: number,
    private environmentName: string
  ) {}
  header(): Html {
    return html`<th scope="col" class="dataset-name-col">Dataset</th>`;
  }
  row(d: HnyTricksDataset, i: number): Html {
    return html`<th scope="row" class="dataset-name-col">${
      d.name
    }<input type="hidden" name="dataset_name_${i}" value="${encodeURIComponent(
      d.name
    )}"/></td>`; // TODO: closing tag is wrong
  }
  footer(): Html {
    return html`<td>
      ${this.countOfDatasets} datasets in ${this.environmentName}
    </td>`;
  }
}

class LinkToSettings implements Column {
  constructor(private environmentUrl: string) {}
  header(): Html {
    return html`<th scope="col">Settings</th>`;
  }
  row(d: HnyTricksDataset): Html {
    const datasetUrl = this.environmentUrl + "datasets/" + d.slug;
    const linkToSettings = datasetUrl + "/overview";
    return html`<td>
      <a href="${linkToSettings}" target="_blank" class="link-symbol">â›­</a>
    </td>`;
  }
  footer(): Html {
    const environmentSettingsUrl = this.environmentUrl + "overview";
    return html`<td>
      <a href="${environmentSettingsUrl}" target="_blank" class="link-symbol">
        â›­
      </a>
    </td>`;
  }
}

const COUNT_QUERY = {
  time_range: 5184000, // last 60 days
  granularity: 0,
  calculations: [
    {
      op: "COUNT",
    },
  ],
  filter_combination: "AND",
  limit: 1000,
};

const CountQueryUrlParams =
  "?query=" + encodeURIComponent(JSON.stringify(COUNT_QUERY));

class LinkToQuery implements Column {
  constructor(private environmentUrl: string) {}
  header(): Html {
    return html`<th scope="col">Query</th>`;
  }
  row(d: HnyTricksDataset): Html {
    const datasetUrl = this.environmentUrl + "datasets/" + d.slug;
    const linkToCountQuery = datasetUrl + CountQueryUrlParams;
    return html`<td>
      <a href="${linkToCountQuery}" target="_blank" class="link-symbol">ðŸ“‰</a>
    </td>`;
  }
  footer(): Html {
    const environmentQueryUrl = this.environmentUrl + CountQueryUrlParams;
    return html`<td>
      <a href="${environmentQueryUrl}" target="_blank" class="link-symbol">
        ðŸ“ˆ
      </a>
    </td>`;
  }
}
