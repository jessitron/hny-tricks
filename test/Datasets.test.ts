import { HnyTricksAuthorization } from "../src/common";
import pretty from "pretty";
import { HnyTricksDataset } from "../src/datasets/dataset_common";
import { DatasetsTable } from "../src/datasets/DatasetsTable";
import { datasetSlugsToDelete } from "../src/datasets/deletion";

expect.addSnapshotSerializer({
  test: (val) => typeof val === "string",
  //@ts-ignore
  print: (val: string) => pretty(val, { ocd: true }), // this works fine
});

const fakeAuth: HnyTricksAuthorization = {
  apiKey: "deadf00t",
  apiKeyId: "what is this",
  keyInfo: {
    type: "configuration",
    environmentType: "e&s",
    region: "local test",
  },
  permissions: {
    canManageDatasets: false,
    canSendEvents: false,
    canManageColumns: false,
  },
  environment: {
    slug: "env_slug",
    name: "my environment",
  },
  team: {
    slug: "team_slug",
    name: "my team name",
  },
};

const twoDaysAgo = new Date(Date.now() - 2 * 24 * 60 * 60 * 1000);
const aYearAgo = new Date(Date.now() - 365 * 24 * 60 * 60 * 1000);
const fakeRecentDataset: HnyTricksDataset = {
  name: "API gateway",
  slug: "api-gateway",
  created: aYearAgo,
  last_written: twoDaysAgo,
};
const fakeExpiredDataset: HnyTricksDataset = {
  name: "ye olde proxy",
  slug: "olde-proxy",
  created: aYearAgo,
  last_written: aYearAgo,
};

describe("DatasetsTable", () => {
  it("outputs the same thing as before, for an empty one", () => {
    const html = DatasetsTable({ datasets: [], auth: fakeAuth });
    expect(html).toMatchSnapshot();
  });
  it("Renders datasets, the same as before", () => {
    const html = DatasetsTable({
      datasets: [fakeRecentDataset, fakeExpiredDataset],
      auth: fakeAuth,
    });
    expect(html).toMatchSnapshot();
  });
});

describe("deleting datasets", () => {
  it("parses the input", () => {
    const input = {
      "dataset-slug-0": "hny-tricks-nodejs",
      "dataset-slug-1": "hny-otel-web-test",
      "dataset-slug-2": "hny-tricks-web",
      "dataset-slug-3": "backend-for-frontend",
      "delete-dataset-3": "on",
      "dataset-slug-4": "unknown_metrics",
      "dataset-slug-5": "phrase-picker-java",
      "dataset-slug-6": "image-picker-java",
      "dataset-slug-7": "meminator-java",
      "dataset-slug-8": "unknown_service",
      "dataset-slug-9": "website",
    };

    const output = datasetSlugsToDelete(input);
    expect(output).toStrictEqual(["backend-for-frontend"]);
  });
});
