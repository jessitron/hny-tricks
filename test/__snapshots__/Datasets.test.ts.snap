// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetsTable Renders datasets, the same as before 1`] = `
<form id="dataset-table-form">
  <table class="dataset-table">
    <thead>
      <tr>
        <th scope="col" class="dataset-name-col">Dataset</th>
        <th scope="col">Settings</th>
        <th scope="col">Query</th>
        <th scope="col">Days Since Last Data</th>
        <th scope="col">Delete?</th>
      </tr>
    </thead>
    <tr>
      <th scope="row" class="dataset-name-col">API gateway</th>
      <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/datasets/api-gateway/overview" target="_blank" class="link-symbol">â›­</a></td>
      <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/datasets/api-gateway?query=%7B%22time_range%22%3A5184000%2C%22granularity%22%3A0%2C%22calculations%22%3A%5B%7B%22op%22%3A%22COUNT%22%7D%5D%2C%22filter_combination%22%3A%22AND%22%2C%22limit%22%3A1000%7D" target="_blank" class="link-symbol">ðŸ“‰</a></td>
      <td>2</td>
      <td><input name="dataset-slug-0" value="api-gateway" type="hidden"><input class="delete-dataset-checkbox" type="checkbox" disabled="true"></td>
    </tr>
    <tr>
      <th scope="row" class="dataset-name-col">ye olde proxy</th>
      <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/datasets/olde-proxy/overview" target="_blank" class="link-symbol">â›­</a></td>
      <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/datasets/olde-proxy?query=%7B%22time_range%22%3A5184000%2C%22granularity%22%3A0%2C%22calculations%22%3A%5B%7B%22op%22%3A%22COUNT%22%7D%5D%2C%22filter_combination%22%3A%22AND%22%2C%22limit%22%3A1000%7D" target="_blank" class="link-symbol">ðŸ“‰</a></td>
      <td>365</td>
      <td><input name="dataset-slug-1" value="olde-proxy" type="hidden"><input class="delete-dataset-checkbox" type="checkbox" name="delete-dataset-1" checked="true"></td>
    </tr>
    <tfoot>
      <tr>
        <td>2 datasets in my environment</td>
        <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/overview" target="_blank" class="link-symbol">â›­</a></td>
        <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/?query=%7B%22time_range%22%3A5184000%2C%22granularity%22%3A0%2C%22calculations%22%3A%5B%7B%22op%22%3A%22COUNT%22%7D%5D%2C%22filter_combination%22%3A%22AND%22%2C%22limit%22%3A1000%7D" target="_blank" class="link-symbol">ðŸ“ˆ</a></td>
        <td>2</td>
        <td><button hx-post="/datasets/delete" hx-target="#dataset-section" hx-include="#auth_data">Delete Old Datasets</button></td>
      </tr>
    </tfoot>
  </table>
</form>
`;

exports[`DatasetsTable outputs the same thing as before, for an empty one 1`] = `
<form id="dataset-table-form">
  <table class="dataset-table">
    <thead>
      <tr>
        <th scope="col" class="dataset-name-col">Dataset</th>
        <th scope="col">Settings</th>
        <th scope="col">Query</th>
        <th scope="col">Days Since Last Data</th>
        <th scope="col">Delete?</th>
      </tr>
    </thead>
    <tfoot>
      <tr>
        <td> datasets in my environment</td>
        <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/overview" target="_blank" class="link-symbol">â›­</a></td>
        <td><a href="http://localhost:3000/test-region/ui/team_slug/environments/env_slug/?query=%7B%22time_range%22%3A5184000%2C%22granularity%22%3A0%2C%22calculations%22%3A%5B%7B%22op%22%3A%22COUNT%22%7D%5D%2C%22filter_combination%22%3A%22AND%22%2C%22limit%22%3A1000%7D" target="_blank" class="link-symbol">ðŸ“ˆ</a></td>
        <td>Infinity</td>
        <td><button hx-post="/datasets/delete" hx-target="#dataset-section" hx-include="#auth_data">Delete Old Datasets</button></td>
      </tr>
    </tfoot>
  </table>
</form>
`;
